planner/core/rule_join_reorder.go

```diff
@@ -107,6 +116,47 @@ func (s *joinReOrderSolver) optimizeRecursive(ctx sessionctx.Context, p LogicalP
 			joinTypes:  joinTypes,
 		}
 
+		joinGroupNum := len(curJoinGroup)
+		var leadingJoinGroup []LogicalPlan
+		if hintInfo != nil {
+			for _, hintTbl := range hintInfo.leadingJoinOrder {
+				for i, joinGroup := range curJoinGroup {
+					tableAlias := extractTableAlias(joinGroup, joinGroup.SelectBlockOffset())
+					if tableAlias == nil {
+						continue
+					}
+					if hintTbl.dbName.L == tableAlias.dbName.L && hintTbl.tblName.L == tableAlias.tblName.L && hintTbl.selectOffset == tableAlias.selectOffset {
+						leadingJoinGroup = append(leadingJoinGroup, joinGroup)
+						curJoinGroup = append(curJoinGroup[:i], curJoinGroup[i+1:]...)
+						break
+					}
+				}
+			}
+			var errMsg string
+			if len(leadingJoinGroup) != len(hintInfo.leadingJoinOrder) {
+				errMsg = "leading hint is inapplicable, check if the leading hint table is valid"
+			} else if joinGroupNum <= ctx.GetSessionVars().TiDBOptJoinReorderThreshold {
+				errMsg = "leading hint is inapplicable for the DP join reorder algorithm"
+			}
+			if len(errMsg) > 0 {
+				curJoinGroup = append(curJoinGroup, leadingJoinGroup...)
+				leadingJoinGroup = nil
+				ctx.GetSessionVars().StmtCtx.AppendWarning(ErrInternal.GenWithStack(errMsg))
+			}
+			if leadingJoinGroup != nil {
+				leadingJoin := leadingJoinGroup[0]
+				leadingJoinGroup = leadingJoinGroup[1:]
+				for len(leadingJoinGroup) > 0 {
+					var usedEdges []*expression.ScalarFunction
+					var joinType JoinType
+					leadingJoin, leadingJoinGroup[0], usedEdges, joinType = baseGroupSolver.checkConnection(leadingJoin, leadingJoinGroup[0])
+					leadingJoin, baseGroupSolver.otherConds = baseGroupSolver.makeJoin(leadingJoin, leadingJoinGroup[0], usedEdges, joinType)
```

Comment:
Should we consider outer join here? For example, in `select /*+ leading(t1, t3) */ * from t1 left join t2 on t1.a=t2.a left join t3 on t2.b=t3.b`, `t1` and `t3` seem unjoinable?